xapian-letor/ranker/listnet_ranker.cc

```diff
@@ -0,0 +1,264 @@
+#include <cmath>
+#include <cstdlib>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <limits>
+#include <sstream>
+#include <vector>
+
+using namespace std;
+using namespace Xapian;
+
+ListNETRanker::ListNETRanker() {
+    // default values
+    learning_rate = 0.0001;
+    iterations = 15;
+    cout << "Initializing ListNETRanker with default parameters" << endl;
+}
+
+ListNETRanker::ListNETRanker(double learn_rate, int num_iterations) {
+    learning_rate = learn_rate;
+    iterations = num_iterations;
+    cout << "Initializing ListNETRanker with Learning Rate: " << learn_rate;
+    cout << ", No. of Iterations: " << iterations << endl;
+}
+
+ListNETRanker::~ListNETRanker() {
+}
+
+static double
+calculateInnerProduct(vector<double> parameters, vector<double> fvals) {
+
+	double inner_product = 0.0;
+
+	for (int i = 0; i < int(fvals.size()); ++i){
+
+		inner_product += parameters[i] * fvals[i];
+	}
+
+	return inner_product;
+}
+
+// From Theorem (6) in paper.
+static vector< vector<double> >
+initializeProbability(vector<FeatureVector> feature_vectors, vector<double> & new_parameters) {
+
+    int list_length = feature_vectors.size();
+
+    // probability distribution, y is ground truth, while z is predict score
+    vector<double> prob_y;
+    vector<double> prob_z;
+    double expsum_y = 0.0;
+    double expsum_z = 0.0;
+
+    for (int i = 0; i < list_length; i++) {
+        expsum_y += exp(feature_vectors[i].get_label());
+        expsum_z += exp(calculateInnerProduct(new_parameters,feature_vectors[i].get_fvals()));
+    }
+
+    for (int i = 0; i < list_length; i++) {
+        prob_y.push_back(exp(feature_vectors[i].get_label())/expsum_y);
+        prob_z.push_back(exp(calculateInnerProduct(new_parameters,feature_vectors[i].get_fvals()))/expsum_z);
+
+    }
+
+    vector< vector<double> > prob;
+    prob.push_back(prob_y);
+    prob.push_back(prob_z);
+
+	return prob;
+}
+
+// Equation (3) in paper.
+static vector<double>
+calculateGradient(vector<FeatureVector> feature_vectors, vector< vector<double> > prob) {
+
+	vector<double> gradient(feature_vectors[0].get_fcount(),0);
+	int list_length = feature_vectors.size();
+
+	vector<double> prob_y(prob[0]);
+	vector<double> prob_z(prob[1]);
+
+	for (int i = 0; i < list_length; i++) {
+		vector<double> fvals = feature_vectors[i].get_fvals();
+
+		for (int k = 0; k < int(fvals.size()); k++){
+
+			double first_term = - prob_y[i] * fvals[k];
+			gradient[k] += first_term;
+
+			double second_term = prob_z[i] * fvals[k];
+			gradient[k] += second_term;
+
+		}
+	}
+
+	return gradient;
+}
+
+static void
+updateParameters(vector<double> & new_parameters, vector<double> gradient, double learning_rate) {
+	int num = new_parameters.size();
+
+	for (int i = 0; i < num; i++){
+		new_parameters[i] -= gradient[i] * learning_rate;
+	}
+
+}
+
+static void
+batchLearning(const std::vector<FeatureVector> & fvv, vector<double> & new_parameters, double learning_rate) {
```

Comment:
Given how short this is, I'd be inclined to inline it in the loop within `train_model()` below?